bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesReferenceProvider.java

```diff
@@ -0,0 +1,189 @@
+
+    @Override
+    public List<Reference> findReferences(Resource resource) {
+        List<Reference> references = new ArrayList<>();
+
+        if (!resource.getPath().startsWith("/content")) {
+            return references;
+        }
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
+        PageManager pageManager = resourceResolver.adaptTo(PageManager.class);
+        if (pageManager == null) {
+            return references;
+        }
+
+        Page page = pageManager.getContainingPage(resource);
+        if (page == null) {
+            return references;
+        } else {
+
+            collectComponentResourceTypes(page.getContentResource());
+
+            if (page.getTemplate() != null) {
+                Resource templateResource = resourceResolver
+                    .getResource(page.getTemplate().getPath());
+                collectComponentResourceTypes(templateResource);
+            }
+
+            if (componentResourceTypes.size() > 0) {
+                references = findSharedComponents(resource);
+            }
+        }
+
+        return references;
+    }
+
+
+    /**
+     * Collect the resourceType of current resource and its children recursively.
+     * Avoided the usage of resource.getResourceType() as it returns the "jcr:primaryType"
+     * when the resource doesn't have "sling:resourceType" property.
+     *
+     * @param resource {@link Resource}
+     */
+    private void collectComponentResourceTypes(Resource resource) {
+        if (resource != null) {
+            String componentResourceType = resource.getValueMap()
+                .get(JcrResourceConstants.SLING_RESOURCE_TYPE_PROPERTY, String.class);
+            if (StringUtils.isNotEmpty(componentResourceType)) {
+                componentResourceTypes.add(componentResourceType);
+            }
+            if (resource.hasChildren()) {
+                Iterator<Resource> children = resource.listChildren();
+                while (children.hasNext()) {
+                    collectComponentResourceTypes(children.next());
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Iterates through the given collection of resourceTypes and finds all the components that has
+     * shared dialog, then adds the root page of the given resource to the list.
+     *
+     * @param resource {@link Resource}
+     * @return references
+     */
+    private List<Reference> findSharedComponents(Resource resource) {
+        List<Reference> references = new ArrayList<>();
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
+        for (String resourceType : componentResourceTypes) {
+
+            Resource componentResource = getComponentResource(resourceType, resourceResolver);
+            if (componentResource == null) {
+                continue;
+            }
+
+            if (!componentResource.getResourceType().equalsIgnoreCase("cq:Component")) {
```

Comment:
Would this be better to check if the `jcr:primaryType` .. while the primary type IS the fall back for a non-existent sling:resourceType, you technically could have a sling:resourceType on a cq:Component for some "weird" use case, and looking for the jcr:primaryType is simply more clear since that's what you actually want to check (i think).

Reply:
Updated the check, and it looks cleaner now. Thanks!